pkg/controller/pxc/status.go

```diff
@@ -23,228 +21,104 @@ const maxStatusesQuantity = 20
+			status: &cr.Status.PXC,
+			spec:   cr.Spec.PXC.PodSpec,
+		},
 	}
 
-	cr.Status.PXC = pxcStatus
-	cr.Status.Host = cr.Name + "-" + "pxc." + cr.Namespace
-	if cr.Status.PXC.Message != "" {
-		cr.Status.Messages = append(cr.Status.Messages, "PXC: "+cr.Status.PXC.Message)
+	cr.Status.HAProxy = api.AppStatus{}
+	if cr.HAProxyEnabled() {
+		apps = append(apps, sfsstatus{
+			app:    statefulset.NewHAProxy(cr),
+			status: &cr.Status.HAProxy,
+			spec:   cr.Spec.HAProxy,
+		})
 	}
 
-	inProgres := false
-
-	if cr.Spec.HAProxy != nil && cr.Spec.HAProxy.Enabled {
-		haProxyStatus, err := r.appStatus(statefulset.NewHAProxy(cr), cr.Spec.HAProxy, cr.Namespace, cr.CompareVersionWith("1.7.0") >= 0)
-		if err != nil {
-			return fmt.Errorf("get haproxy status: %v", err)
-		}
-		haProxyStatus.Version = cr.Status.HAProxy.Version
-
-		if haProxyStatus.Status != cr.Status.HAProxy.Status {
-			if haProxyStatus.Status == api.AppStateReady {
-				clusterCondition = api.ClusterCondition{
-					Status:             api.ConditionTrue,
-					Type:               api.ClusterHAProxyReady,
-					LastTransitionTime: metav1.NewTime(time.Now()),
-				}
-			}
-
-			if haProxyStatus.Status == api.AppStateError {
-				clusterCondition = api.ClusterCondition{
-					Status:             api.ConditionTrue,
-					Message:            "HAProxy:" + haProxyStatus.Message,
-					Reason:             "ErrorHAProxy",
-					Type:               api.ClusterError,
-					LastTransitionTime: metav1.NewTime(time.Now()),
-				}
-			}
-		}
-
-		cr.Status.HAProxy = haProxyStatus
-
-		cr.Status.Host = cr.Name + "-" + "haproxy." + cr.Namespace
-		if cr.Spec.HAProxy.ServiceType == corev1.ServiceTypeLoadBalancer {
-			svc := &corev1.Service{}
-			err := r.client.Get(context.TODO(),
-				types.NamespacedName{
-					Namespace: cr.Namespace,
-					Name:      cr.Name + "-" + "haproxy",
-				}, svc)
-			if err != nil {
-				return errors.Wrap(err, "get haproxy service")
-			}
-			for _, i := range svc.Status.LoadBalancer.Ingress {
-				cr.Status.Host = i.IP
-				if len(i.Hostname) > 0 {
-					cr.Status.Host = i.Hostname
-				}
-			}
-		}
-		if cr.Status.HAProxy.Message != "" {
-			cr.Status.Messages = append(cr.Status.Messages, "HAProxy: "+cr.Status.HAProxy.Message)
-		}
-		inProgres, err = r.upgradeInProgress(cr, "haproxy")
-		if err != nil {
-			return fmt.Errorf("check haproxy upgrade progress: %v", err)
-		}
-	} else {
-		cr.Status.HAProxy = api.AppStatus{}
+	cr.Status.ProxySQL = api.AppStatus{}
+	if cr.ProxySQLEnabled() {
+		apps = append(apps, sfsstatus{
+			app:    statefulset.NewProxy(cr),
+			status: &cr.Status.ProxySQL,
+			spec:   cr.Spec.ProxySQL,
+		})
 	}
 
-	if cr.Spec.ProxySQL != nil && cr.Spec.ProxySQL.Enabled {
-		proxyStatus, err := r.appStatus(statefulset.NewProxy(cr), cr.Spec.ProxySQL, cr.Namespace, cr.CompareVersionWith("1.7.0") >= 0)
+	inProgress := false
+
+	for _, a := range apps {
+		*a.status, cr.Status.Host, err = r.componentStatus(a.app, cr, a.spec, *a.status)
```

Comment:
a bit hackish calculation of cr.Status.Host, but up to @dAdAbird

Reply:
see https://github.com/percona/percona-xtradb-cluster-operator/commit/337d9d567bbad589cfd8d851d66960f2a55d8ac0